File: python/semanticabi/steps/AbiSchema.py
from __future__ import annotations
from typing import List, Dict, Callable

from semanticabi.abi.InvalidAbiException import InvalidAbiException
from semanticabi.common.column.DatasetColumn import DatasetColumn
from semanticabi.common.column.RenamedColumn import RenamedColumn


class AbiSchema:
    """
    The column schema from parsing a semantic abi
    """

    _columns: List[DatasetColumn]
    _column_indices: Dict[str, int]

    def __init__(self, columns: List[DatasetColumn] = None):
        if columns is None:
            columns = []
        self._columns = columns
        self._column_indices = {column.name: index for index, column in enumerate(columns)}

    def columns(self) -> List[DatasetColumn]:
        return self._columns

    def column(self, name: str) -> DatasetColumn:
        return self._columns[self._column_indices[name]]

    def has_column(self, name: str) -> bool:
        return name in self._column_indices

    def with_columns(self, columns: List[DatasetColumn], allow_overwrite: bool = False) -> AbiSchema:
        """
        Append the list of columns to the existing columns
        """
        new_columns: List[DatasetColumn] = self._columns.copy()
        for column in columns:
            if column.name in self._column_indices and not allow_overwrite:
                raise InvalidAbiException(f'Column \'{column.name}\' already exists in schema')

            if column.name in self._column_indices:
                new_columns[self._column_indices[column.name]] = column
            else:
                new_columns.append(column)

        return AbiSchema(new_columns)

    def append_schema_with_rename(self, schema: AbiSchema, column_rename_fn: Callable[[str], str]) -> AbiSchema:
        """
        Append the schema to this schema while renaming each column using the given rename function
        """
        new_columns: List[DatasetColumn] = self._columns.copy()
        for column in schema.columns():
            new_column_name = column_rename_fn(column.name)
            if new_column_name in self._column_indices:
                raise InvalidAbiException(f'Column \'{new_column_name}\' already exists in schema')

            new_columns.append(RenamedColumn(column, new_column_name))

        return AbiSchema(new_columns)

    def __eq__(self, other: AbiSchema) -> bool:
        return self._columns == other._columns

    def __ne__(self, other: AbiSchema) -> bool:
        return not self.__eq__(other)

    def __hash__(self):
        return hash(self._columns)

    def __str__(self):
        return f'AbiSchema({",".join([str(column) for column in self._columns])})'

File: js/src/semanticabi/transform/steps/AbiSchema.ts
import {DatasetColumn} from "semanticabi/transform/column/DatasetColumn";
import {OrderedMap} from "common/OrderedMap";
import {Optional} from "common/Optional";
import {SemanticAbiExecutionError} from "semanticabi/abi/item/semantic/SemanticAbiExecutionError";

/**
 * @author zuyezheng
 */
export class AbiSchema {

    private readonly _columns: OrderedMap<string, DatasetColumn>;

    constructor(columns: DatasetColumn[]) {
        this._columns = OrderedMap.fromKeyed(columns, c => c.name);
    }

    get columns(): DatasetColumn[] {
        return this._columns.values;
    }

    column(name: string): Optional<DatasetColumn> {
        return this._columns.get(name);
    }

    hasColumn(name: string): boolean {
        return this._columns.has(name);
    }

    /**
     * Append the list of columns to the existing columns.
     */
    withColumns(columns: DatasetColumn[], allowDupes: boolean = false): AbiSchema {
        if (allowDupes) {
            // make sure if the column is a dupe, it is of the same type
            Optional.of(columns.find(c =>
                // see if the column exists
                this._columns.get(c.name)
                    // if it does, return the column if there is a type mismatch so we can throw with it
                    .map(existingC => c.dataType !== existingC.dataType)
                    // all good
                    .getOr(false)
            )).forEach(c => {
                throw new SemanticAbiExecutionError(`Column with name '${c.name}' already exists and is a different type.`);
            });
        } else {
            // strictly check for dupes
            Optional.of(columns.find(c => this.hasColumn(c.name)))
                .forEach(c => {
                    throw new SemanticAbiExecutionError(`Column with name '${c.name}' already exists.`);
                });
        }

        return new AbiSchema(this.columns.concat(columns));
    }

    /**
     * Merge 2 schemas without allowing for duplicate column names, returning a new merged.
     */
    mergeSchema(other: AbiSchema, rename: (name: string) => string, allowDupes: boolean = false): AbiSchema {
        return this.withColumns(other.columns.map(c => c.withName(rename(c.name))), allowDupes);
    }

}